osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs

```diff
@@ -0,0 +1,399 @@
+                        }
+                        else
+                        {
+                            // New segment with the same type - duplicate the control point
+                            writer.Write(FormattableString.Invariant($"{positionData.X + point.Position.Value.X}:{positionData.Y + point.Position.Value.Y}|"));
+                        }
+                    }
+
+                    if (i != 0)
+                    {
+                        writer.Write(FormattableString.Invariant($"{positionData.X + point.Position.Value.X}:{positionData.Y + point.Position.Value.Y}"));
+                        writer.Write(i != curveData.Path.ControlPoints.Count - 1 ? "|" : ",");
+                    }
+                }
+
+                writer.Write(FormattableString.Invariant($"{curveData.RepeatCount + 1},"));
+                writer.Write(FormattableString.Invariant($"{curveData.Path.Distance},"));
+
+                for (int i = 0; i < curveData.NodeSamples.Count; i++)
+                {
+                    writer.Write(FormattableString.Invariant($"{(int)toLegacyHitSoundType(curveData.NodeSamples[i])}"));
+                    writer.Write(i != curveData.NodeSamples.Count - 1 ? "|" : ",");
+                }
+
+                for (int i = 0; i < curveData.NodeSamples.Count; i++)
+                {
+                    writer.Write(getSampleBank(curveData.NodeSamples[i], true));
+                    writer.Write(i != curveData.NodeSamples.Count - 1 ? "|" : ",");
+                }
+            }
+            else if (hitObject is IHasEndTime endTimeData)
+                writer.Write(FormattableString.Invariant($"{endTimeData.EndTime},"));
+
+            writer.Write(hitObject is IHasCurve
+                ? getSampleBank(hitObject.Samples, zeroBanks: true)
+                : getSampleBank(hitObject.Samples));
+
+            writer.Write(Environment.NewLine);
+        }
+
+        private void handleTaikoHitObject(TextWriter writer, HitObject hitObject)
+        {
+        }
+
+        private void handleCatchHitObject(TextWriter writer, HitObject hitObject)
+        {
+        }
+
+        private void handleManiaHitObject(TextWriter writer, HitObject hitObject)
+        {
+        }
+
+        private string getSampleBank(IList<HitSampleInfo> samples, bool banksOnly = false, bool zeroBanks = false)
+        {
+            LegacySampleBank normalBank = toLegacySampleBank(samples.FirstOrDefault(s => s.Name == HitSampleInfo.HIT_NORMAL)?.Bank);
+            LegacySampleBank addBank = toLegacySampleBank(samples.FirstOrDefault(s => !string.IsNullOrEmpty(s.Name) && s.Name != HitSampleInfo.HIT_NORMAL)?.Bank);
+
+            string customSampleBank = toLegacyCustomSampleBank(samples.FirstOrDefault()?.Suffix);
+            string sampleFilename = samples.FirstOrDefault(s => string.IsNullOrEmpty(s.Name))?.LookupNames.First() ?? string.Empty;
+
+            int volume = samples.First().Volume;
```

Comment:
Maybe these should be `Single(OrDefault)`s?

Reply:
I've made the change where possible, although this is not the case for every `First*()` in this method.